### PR TITLE
Switch to using RPMs instead of the SDK.

### PR DESCRIPTION
**Pull request by aslicerh**
- This includes some temporary changes specific to the 2.0 preview
- This also addresses issue #87, which makes the runtime container run as the default user
- Corrected the runtime container name from 'dotnet-20-rhel7-runtime' to 'dotnet-20-runtime-rhel'
- Removed the package cache warmup from the build image.

On a side note, I know it is preferred to cat-until-EOF into files instead of repeated echos, but I tried a few different forms of syntax and docker does not seem to like that being inside of a dockerfile. If anyone knows how to pull that off, please let me know.

### CODE REVIEW

File: 2.0/runtime/Dockerfile.rhel7 [hunk 1]
```diff
@@ -33,16 +38,38 @@ ENV DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1
 ## the container port.
 ENV ASPNETCORE_URLS=http://*:8080
 
-RUN yum install -y --setopt=tsflags=nodocs nss_wrapper tar && \
+# Each language image can have 'contrib' a directory with extra files needed to
+# run and build the applications.
+COPY ./contrib/ /opt/app-root
+
+# TODO: This is needed for the 2.0 preview. It should be removed for the final release.
```
Comment (omajid) [inline 2.0/runtime/Dockerfile.rhel7]: Actually, this should not be needed as soon as the RPMs get published into the beta channel

File: 2.0/runtime/Dockerfile.rhel7 [hunk 2]
```diff
@@ -33,16 +38,38 @@ ENV DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1
...
+    echo "enabled=1" >> rh-dotnet20.repo && \
+    echo "gpgcheck=0" >> rh-dotnet20.repo && \
+    cp rh-dotnet20.repo /etc/yum.repos.d/
+
+RUN yum install --disablerepo=\* --enablerepo=rhel-7-server-rpms -y yum-utils && \
+    yum-config-manager --enable rhel-server-rhscl-7-rpms && \
+    yum-config-manager --enable rhel-7-server-dotnet-rpms && \
+    yum-config-manager --enable rhel-7-server-rpms && \
+    yum-config-manager --enable rhel-7-server-ose-3.2-rpms && \
+    yum install -y --setopt=tsflags=nodocs rh-dotnet20-dotnet-runtime-2.0 nss_wrapper tar libicu && \
```
Comment (omajid) [inline 2.0/runtime/Dockerfile.rhel7]: `libicu` should now get pulled in automatically. When you get some time, can you try and confirm things work when it is removed?

**Comment by omajid** [comment]
👍

**Comment by jerboaa** [comment]
I'd suggest to source `generate_container_user` in [DOTNET_RUN_SCRIPT](https://github.com/redhat-developer/s2i-dotnetcore/blob/2.0-dev/2.0/build/s2i/bin/assemble#L83) and also to refresh test binaries in `runtime/test` so as to have the updated s2i_run script including the sourced `generate_container_user` line.

**Comment by jerboaa** [comment]
[cibuild]

**Comment by jerboaa** [comment]
For some reason the CI builder does not seem to build the runtime image from scratch. I don't know why. Could you please add this at the very top to `build.sh`:

     if [ "${DEBUG}" == "true"]; then
       set -x
     fi

**Comment by jerboaa** [comment]
@aslicerh Nice work!

File: 2.0/build/README.md [hunk 3]
```diff
@@ -103,6 +103,11 @@ a `.s2i/environment` file inside your source code repository.
     Used to specify the list of NuGet package sources used during the restore operation. This overrides 
     all of the sources specified in the NuGet.config file.
 
+    **NOTE:** For the 2.0 preview, some sources will be added automatically. Specifically, these are:
+    * https://api.nuget.org/v3/index.json
+    * https://dotnet.myget.org/F/dotnet-core/api/v2
+    * https://dotnet.myget.org/F/dotnet-2-0-0-preview2-final/api/v2
+
```
Comment (tmds) [inline 2.0/build/README.md]: The workaround in the readme is sufficient, there is no need to add this in the README. end-users don't need to worry about this. Released versions (including beta/preview) have their dependencies available on nuget.org.

File: 2.0/runtime/Dockerfile.rhel7 [hunk 4]
```diff
@@ -11,13 +11,18 @@ ENV LANG=C.UTF-8 \
     HOME=/opt/app-root \
     PATH=/opt/app-root/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
 
+# Certain versions of docker restrict syscalls that will cause liblttng to fail
+# and dotnet to abort. This flag will prevent that.
+# For more information, see this bug: https://github.com/dotnet/cli/issues/1582
+ENV LTTNG_UST_REGISTER_TIMEOUT=0
```
Comment (tmds) [inline 2.0/runtime/Dockerfile.rhel7]: Do we consider Docker 1.10 a supported version? If not, please remove this.

File: 2.0/runtime/test/run [hunk 5]
```diff
@@ -398,10 +398,12 @@ create_runtime_app() {
 
 cat << EOF >"${TMPDIR}/Dockerfile"
 FROM ${IMAGE_NAME}
+USER 0
```
Comment (tmds) [inline 2.0/runtime/test/run]: Why do we need to switch user? Isn't /opt/app-root owned by 1001?

Comment (aslicerh) [inline 2.0/runtime/test/run]: This is a good point. I don't need to change the user, I need to copy the chown lines from the build dockerfile into the runtime dockerfile.

File: 2.0/runtime/Dockerfile.rhel7 (hunk 4 above)
Comment (jerboaa) [inline 2.0/runtime/Dockerfile.rhel7]: It looks like the rhel-7-server-extras-rpms channel which has `docker` and `docker-latest` has versions `1.12.6-16.el7` and `1.13.1-4.el7` respectively. Might not be 100% accurate, though.

Comment (aslicerh) [inline 2.0/runtime/Dockerfile.rhel7]: `1.12` is the latest standard docker version in rhel7 (though I think that is a recent update). I'm trying to get confirmation now on how far back we support docker though.

**Comment by jerboaa** [comment]
[cibuild]

File: 2.0/runtime/Dockerfile.rhel7 (hunk 4 above)
Comment (aslicerh) [inline 2.0/runtime/Dockerfile.rhel7]: Ok, after some discussion I think I can remove this.

File: 2.0/build/Dockerfile.rhel7 [hunk 6]
```diff
@@ -27,61 +27,42 @@ COPY ./root/usr/bin /usr/bin
...
+# Create the NuGet file necessary for the preview
+RUN cd /opt/app-root/src && \
+    echo "<?xml version=\"1.0\" encoding=\"utf-8\"?>" > NuGet.config && \
+    echo "<configuration>" >> NuGet.config && \
+    echo "  <packageSources>" >> NuGet.config && \
+    echo "    <add key=\"api-nuget\" value=\"https://dotnet.myget.org/F/dotnet-core/api/v2\" />" >> NuGet.config && \
+    echo "    <add key=\"dotnet-myget\" value=\"https://dotnet.myget.org/F/dotnet-core/api/v2\" />" >> NuGet.config && \
+    echo "    <add key=\"dotnet-preview-myget\" value=\"https://dotnet.myget.org/F/dotnet-2-0-0-preview2-final/api/v2\" />" >> NuGet.config && \
+    echo "  </packageSources>" >> NuGet.config && \
+    echo "</configuration>" >> NuGet.config
```
Comment (tmds) [inline 2.0/build/Dockerfile.rhel7]: This can be removed, Microsoft has uploaded preview2 packages to nuget.org.

File: 2.0/build/s2i/bin/assemble [hunk 7]
```diff
@@ -53,8 +53,9 @@ else
 fi
 
 # Build nuget sources list for when doing the restore
+# TODO: These are necessary for the 2.0 preview. They should be removed for the final release.
 # TODO: initialize RESTORE_OPTIONS to empty string
-RESTORE_OPTIONS=""
+RESTORE_OPTIONS="--source https://api.nuget.org/v3/index.json --source https://dotnet.myget.org/F/dotnet-core/api/v2 --source https://dotnet.myget.org/F/dotnet-2-0-0-preview2-final/api/v2"
```
Comment (tmds) [inline 2.0/build/s2i/bin/assemble]: This can be removed, Microsoft has uploaded preview2 packages to nuget.org.

**Comment by jerboaa** [comment]
I'm OK to merge this, even though CI cannot build it properly yet. Modulo @tmds requested changes.

File: 2.0/build/Dockerfile.rhel7 [hunk 8]
```diff
@@ -27,61 +27,42 @@ COPY ./root/usr/bin /usr/bin
 # Copy the S2I scripts from the specific language image to $STI_SCRIPTS_PATH.
 COPY ./s2i/bin/ /usr/libexec/s2i
 
-# Each language image can have 'contrib' a directory with extra files needed to
-# run and build the applications.
-COPY ./contrib/ /opt/app-root
+# Switch back to root for changing dir ownership/permissions
```
Comment (tmds) [inline 2.0/build/Dockerfile.rhel7]: remove 'for changing dir ownership/permissions'

File: 2.0/build/s2i/bin/assemble [hunk 9]
```diff
@@ -82,6 +83,7 @@ fi
 # Create run script in publish folder
 cat << EOF >"$DOTNET_RUN_SCRIPT"
 #!/bin/bash
+source /opt/app-root/etc/generate_container_user
```
Comment (tmds) [inline 2.0/build/s2i/bin/assemble]: We no longer need to call this in `s2i/bin/run` then?

**Comment by jerboaa** [comment]
[cibuild]

**Comment by jerboaa** [comment]
[cibuild]

**Pull request merged**